### ISSUE TITLE
[ENH] conditional forecaster multiplexing by instance property, e.g., slow or fast moving goods

### ISSUE THREAD

**Issue by davidgilbertson** [sktime/sktime#4818]
**Is your feature request related to a problem? Please describe.**
Here is the logic that I would like to use with sktime:
* I have some panel data, where the data is grouped into instances by 'store' and 'product', and the target value is 'Sales'
* I have a special rule that does a different type of forecast based on the product and past sales (it predicts all zeros based on a heuristic that predicts if a store no longer stocks a product)

Now, I have my custom forecaster working well, and its `fit` method is called with `X` for each instance, but there's no way for me to tell which instance I'm currently looking at (that is, which store/product I'm looking at) so I can't apply any product-specific logic.

**Describe the solution you'd like**
I've looked at the code and can't see how it could be done easily. In this code, `group_name` would need to be passed through to the estimator method, but of course that would error if it's not expected. You could `inspect` the func to see if it's expected.
https://github.com/sktime/sktime/blob/f94a4bf4ed6e7ec1520496ec00ae987c57cc9703/sktime/datatypes/_vectorize.py#L572-L584

**Describe alternatives you've considered**
A few workarounds that work just fine:
* Add a column to X `IsProductSomething` so the information is in X
* Put all the logic for whether or not to do a particular type of forecast outside the sktime pipeline, so have a boolean column for `JustPredictZero` and read that in the forecaster.

These both spread out the logic of conditional forecasting which isn't ideal.

I could also put the Product column in X (in addition to being in the index), but it's a string, so then I need to mess around with categories/integer representations, or drop the column before passing to `fit`, and it needs a different name (else it clashes with the index) so that approach is a bit of a mess.

**Additional context**
This is probably quite a niche problem, I'd be perfectly happy for you to close this. But if at some point the opportunity arises to make this happen that would be neat.

**Comment by fkiraly** [comment]
Have you seen `HierarchyEnsembleForecaster` (in `sktime.forecasting.compose`) - this might solve your problem, it allows to apply different forecasters per product or product group.

The `VectorizedDF` that you are looking at just does a naive loop over hierarchy levels if the estimator itself is not genuinely hierarchical (broadcast aka apply-by) - you wouldn't expect (or want) any estimator logic there.

**Comment by fkiraly** [comment]
FYI @VyomkeshVyas

**Comment by davidgilbertson** [comment]
I hadn't seen `HierarchyEnsembleForecaster`. It's close to what I want, but I've made my own `SwitchForecaster` that takes a dictionary of forecasters, and a selector function that decides which of several forecasters to use (returns the appropriate key):
```py
SwitchForecaster(
    estimators={
        "PredictZero": PredictZero(),
        "LagYIntoX": LagYIntoX(
            estimator=HistGradientBoostingRegressor(),
            num_lags=2,
        ),
    },
    selector=lambda X, y: "PredictZero" if y.squeeze().tail(30).sum() < 4 and not X.IsSchoolSupplies[0] else "LagYIntoX",
)
```

**Comment by fkiraly** [comment]
`SwitchForecaster` = `MultiplexForecaster` with lambda expressions as condition?

Interesting, your selector can take lambda expressions and is hence more elaborate, it's a conditional switch!

Mind contributing this? This is a very exciting component!

I've tried sth similar with the `TransformIf`, but the condition is on parameters rather than on data, and currently afaik there is no generic data-dependent switch.

Small note: lambda expressions are not pickleable and hence are non-compliant with the save/load part of the extender contract (you might get into trouble if you try to serialize for deployment).

A workaround that I've used before is allow strings and then use `eval` internally if it's a string - strings are pickleable.

**Comment by davidgilbertson** [comment]
I figured I'd run with it for a while and see what bugs shake out. I think it might be fiddly to use with something like grid search but I haven't got that far yet.

Here's the code though:
```py
class SwitchForecaster(BaseForecaster):
    _tags = {
        "y_inner_mtype": "pd.DataFrame",
        "X_inner_mtype": "pd.DataFrame",
    }

    def __init__(self, estimators: Mapping, selector: Callable):
        super().__init__()
        self.estimators = estimators
        self.estimator = None
        self.selector = selector

    def _fit(self, y, X=None, fh=None):
        est_key = self.selector(X, y)
        self.estimator = self.estimators[est_key]
        self.estimator.fit(y=y, X=X, fh=fh)
        return self

    def _predict(self, fh, X=None):
        return self.estimator.predict(fh=fh, X=X)
```

Yep I'm aware of the lambda/pickle issue (thanks for the heads up though), in reality I'd use a function to keep the pipeline definition neat. Are you aware of cloudpickle? It handles lambdas and I think is a drop-in replacement for pickle.

**Comment by davidgilbertson** [comment]
Got it working with grid search by making the estimator options arguments:
```py
SwitchForecaster(
    predict_zero=PredictZero(),
    lag_y_into_x=LagYIntoX(
        estimator=HistGradientBoostingRegressor(
            categorical_features=cat_cols,
            early_stopping=True,
        ),
        num_lags=2,
    ),
    selector=lambda X, y: "predict_zero" if y.squeeze().tail(30).sum() < 4 and not X.IsSchoolSupplies[0] else "lag_y_into_x",
)
```

And some faffing about to make the clone process work by adding to `get_param_names`:
```py
class SwitchForecaster(BaseForecaster):
    _tags = {
        "y_inner_mtype": "pd.DataFrame",
        "X_inner_mtype": "pd.DataFrame",
    }

    def __init__(self, selector: Callable, **kwargs):
        super().__init__()
        for key, val in kwargs.items():
            setattr(self, key, val)
        self.kwargs = kwargs
        self.estimator = None
        self.selector = selector

    def _fit(self, y, X=None, fh=None):
        est_key = self.selector(X, y)
        self.estimator = getattr(self, est_key)
        self.estimator.fit(y=y, X=X, fh=fh)
        return self

    def _predict(self, fh, X=None):
        return self.estimator.predict(fh=fh, X=X)

    def get_param_names(self):
        param_names = super().get_param_names()
        param_names.extend(self.kwargs.keys())
        return param_names
```

**Comment by fkiraly** [comment]
Looks nice, although you should avoid overrides of public methods like `get_param_names`.

The clean way would be in line with the `scikit-learn` interface conventions, where the set of parameters cannot change on the top level.

If you are interested in an interface compliant implementation, have a look at `MultiplexForecaster`.

**Comment by tpvasconcelos** [comment]
I think I achieved something similar using the `MultiplexForecaster` and `PluginParamsForecaster` primitives, and defining a `ForecasterSelector` parameter fitter:

```python
class ForecasterSelector(BaseParamFitter):
    """Forecaster selector base class.

    This class is used to select a forecaster based on the input data.
    The selected forecaster is stored in the `selected_forecaster_` 
    attribute, which can be then passed down to the 
    ``MultiplexForecaster`` via ``PluginParamsForecaster``.

    Args:
        selector:
            A callable that takes the input time series and returns the 
            name of the forecaster to select.

    """

    def __init__(self, selector: Callable[[pd.DataFrame], str]):
        super().__init__()
        self.selector = selector

    def _fit(self, X: pd.DataFrame) -> BaseParamFitter:
        self.selected_forecaster_ = self.selector(X)
        return self
```

So the equivalent to @davidgilbertson's example would be:

```python
PluginParamsForecaster(
    param_est=ForecasterSelector(
        lambda X: "PredictZero" if X.squeeze().tail(30).sum() < 4 and not X.IsSchoolSupplies[0] else "LagYIntoX"
    ),
    forecaster=MultiplexForecaster(
        [
            ("PredictZero", PredictZero()),
            ("LagYIntoX", LagYIntoX(...)),
        ]
    ),
)
```

**Comment by fkiraly** [comment]
Interesting!

Would you like to contribute this, @tpvasconcelos?
I was not thinking about using parameter fitters/setters, very interesting.

It also seems like a common enough use case, you are already the second person posting about this feature gap, which typically means there are hundreds of others who may like to have it (but are too shy to post 😄 )

**Comment by fkiraly** [comment]
comments on `ForecasterSelector` - that's basically a conditional parameter selector, nothing requires the selected parameter to be a forecaster, no?

**Comment by tpvasconcelos** [comment]
@fkiraly Happy to contribute this. I think I like your `ParamSelector` suggestion. Do you mean something like this:

```python
class ParamSelector(BaseParamFitter):

    def __init__(self, param: str, selector: Callable[[pd.DataFrame], str]):
        super().__init__()
        self.param = param
        self.selector = selector

    def _fit(self, X: pd.DataFrame) -> BaseParamFitter:
        setattr(self, self.param, self.selector(X))
        return self
```

and then:

```python
PluginParamsForecaster(
    param_est=ParamSelector(
        param="selected_forecaster_",
        selector=lambda X: "PredictZero" if X.squeeze().tail(30).sum() < 4 and not X.IsSchoolSupplies[0] else "LagYIntoX",
    ),
    forecaster=MultiplexForecaster(
        [
            ("PredictZero", PredictZero()),
            ("LagYIntoX", LagYIntoX(...)),
        ]
    ),
)
```

Looks nice and pretty generalisable to me! 🚀

**Comment by fkiraly** [comment]
yes, exactly!

In addition, nice syntactic sugar would be to make the `*` dunder on param estimator plus forecaster to default to `PluginParamsForecaster`, but that's a minor thing imo.

**Comment by tpvasconcelos** [comment]
@fkiraly I'm currently working on a draft version of this and taking inspiration from the existing `FunctionTransformer`. One thing that came to mind is that all sktime estimators should have default values for the constructor parameters. Im not sure what a reasonable default argument would be for the `selector` callable (I'm calling it `func` in the draft implementation to keep the same _language_ used in the existing `FunctionTransformer`). For the `FunctionTransformer` it defaults to an identity function which makes sense but in this case, defaulting to a simple `lambda X: None` seems unreasonable. Thoughts?

**Comment by tpvasconcelos** [comment]
Same applies for the `param` parameter. Here `None` doesn't work since it will have to be a string value.

**Comment by tpvasconcelos** [comment]
Sorry... after looking around the codebase for a bit I noticed that some estimator do not have any default values for some parameters. For instance `FixedParams` does not have a default value for `param_dict`. I will move forward making both these parameters mandatory then

**Comment by fkiraly** [comment]
> One thing that came to mind is that all sktime estimators should have default values for the constructor parameters

That is, I think, not an sklearn convnetion, and not an sktime one either. Or, am I wrong?

**Comment by tpvasconcelos** [comment]
<img width="1224" alt="image" src="https://github.com/sktime/sktime/assets/17701527/daaf39d6-f066-4a79-85e4-29fd4c214d48">

https://scikit-learn.org/stable/developers/develop.html#:~:text=The%20arguments%20accepted%20by%20__init__%20should%20all%20be%20keyword%20arguments%20with%20a%20default%20value.%20In%20other%20words%2C%20a%20user%20should%20be%20able%20to%20instantiate%20an%20estimator%20without%20passing%20any%20arguments%20to%20it.

**Comment by fkiraly** [comment]
hm, that's interesting. You are clearly right, and I looked at how far it goes back, and indeed it is already like that since years. I am surprised that I forgot about this, but it's been a long time since sktime started.

So I now assume we dropped that interface assumption for a good reason, possibly since in some cases there are no good defaults. For testing, `get_test_params` always produces complete parameter sets.
